### PR TITLE
overlord/snapstate: prefer a smaller corner case for doing the wrong thing

### PR DESCRIPTION
**Pull request by chipaca**
Currently, if a snap is installed with any flags, refreshed
(preserving those flags), and reverted, the flags are lost.

This means that a classic snap suddenly finds itself confined in
strict mode (ditto for devmode; ditto again for jailmode). The current
reply to this was "revert needs explicit flags", which is surprising
to most users. The reasoning being that a snap can be classic at
revision N, and strict at revision M, and as flags are per snap and
not per revision we can't make a sensible decision.

> The  right fix is to make flags per revision, obviously (in retrospect)

While we find time to implement the right fix, this PR changes the
behaviour so that explicit flags are needed to _change_ the flags that
the snap is installed with; if nothing is specified, flags are
preserved.

This will break the aforementioned flip-flopping snaps, but preserve
the apparently more common case. So the breakage is moved from a
bigger corner case to a smaller one, which is a win.

This builds on #3940.

### CODE REVIEW

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -1372,6 +1372,17 @@ func RevertToRevision(st *state.State, name string, rev snap.Revision, flags Fla
 		return nil, err
 	}
 	flags.Revert = true
+	if !(flags.JailMode || flags.DevMode || flags.Classic) {
```
Comment (zyga) [inline overlord/snapstate/snapstate.go]: How about adding a comment here:
```
// TODO: revert flags from the revision `rev` once state tracks per-revision flags.
```

**Comment by mvo5** [comment]
Test failures in spread look related?

**Comment by chipaca** [comment]
yes, seems i broke a spread test. will fix.

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -1372,6 +1372,17 @@ func RevertToRevision(st *state.State, name string, rev snap.Revision, flags Fla
...
+		if snapst.Flags.DevMode {
+			flags.DevMode = true
+		}
+		if snapst.Flags.JailMode {
+			flags.JailMode = true
+		}
+		if snapst.Flags.Classic {
+			flags.Classic = true
```
Comment (niemeyer) [inline overlord/snapstate/snapstate.go]: Should we at least see if the snap is classic before doing that?

Comment (chipaca) [reply overlord/snapstate/snapstate.go]: we don't have the snap, at this point

**Comment by zyga** [comment]
The test failures on forced-devmode systems look real:
```
snap revert --jailmode test-snapd-tools
error: cannot revert "test-snapd-tools": this system cannot honour the jailmode flag
```

**Pull request merged**